File: javaLoginCore/src/main/java/com/hamdan/forzenbook/java/login/core/data/repository/LoginRepositoryImpl.java
package com.hamdan.forzenbook.java.login.core.data.repository;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.hamdan.forzenbook.java.login.core.data.database.LoginDao;
import com.hamdan.forzenbook.java.login.core.data.network.LoginResponse;
import com.hamdan.forzenbook.java.login.core.data.network.LoginService;

import retrofit2.Response;

public class LoginRepositoryImpl implements LoginRepository {
    final private LoginDao dao;
    final private LoginService service;

    public LoginRepositoryImpl(LoginDao dao, LoginService service) {
        this.dao = dao;
        this.service = service;
    }

    @Override
    public void requestValidation(String email) {
        service.requestValidation(email);
    }

    @Override
    @NonNull
    public User getToken(@Nullable String email, @Nullable String code) throws NullTokenException, FailTokenRetrievalException {
        if (email == null && code == null) return getTokenFromDatabase();
        else return getTokenFromNetwork(email, code);
    }


    @NonNull
    private User getTokenFromDatabase() {
        if (dao.getToken() == null)
            return User.NOT_LOGGED_IN;
        else return User.LOGGED_IN;
    }

    @NonNull
    private User getTokenFromNetwork(String email, String code) throws FailTokenRetrievalException, NullTokenException {
        Response<LoginResponse> response = service.getToken(email, code);
        if (response.isSuccessful()) {
            if (response.body() == null) {
                throw new FailTokenRetrievalException("Failed to get Token");
            } else {
                LoginResponse body = response.body();
                if (body.getToken() == null) {
                    throw new NullTokenException("null token");
                } else return User.LOGGED_IN;
            }
        } else throw new FailTokenRetrievalException("Failed to get Token");
    }
}

File: loginCore/src/main/java/com/hamdan/forzenbook/login/core/data/repository/LoginRepositoryImpl.kt
package com.hamdan.forzenbook.login.core.data.repository

import android.content.Context
import android.util.Log
import com.hamdan.forzenbook.core.FailTokenRetrievalException
import com.hamdan.forzenbook.core.GlobalConstants
import com.hamdan.forzenbook.core.NullTokenException
import com.hamdan.forzenbook.core.getToken
import com.hamdan.forzenbook.core.removeToken
import com.hamdan.forzenbook.login.core.data.network.LoginService

class LoginRepositoryImpl(
  private val service: LoginService,
  private val context: Context,
) : LoginRepository {
  override suspend fun requestValidation(email: String) {
    service.requestValidation(email)
  }

  override suspend fun getToken(email: String?, code: String?) {
    if (email.isNullOrEmpty() && code.isNullOrEmpty()) getToken(context)
      ?: throw NullTokenException()
    else email?.let { code?.let { getTokenFromNetwork(email, code) } }
  }

  private suspend fun getTokenFromNetwork(email: String, code: String) {
    val response = service.getToken(email, code)
    if (response.isSuccessful) {
      val body = response.body() ?: throw FailTokenRetrievalException("Failed to get Token")
      body.token ?: throw NullTokenException("null token")
      context.getSharedPreferences(
        GlobalConstants.TOKEN_PREFERENCE_LOCATION,
        Context.MODE_PRIVATE
      ).edit().apply {
        putString(GlobalConstants.TOKEN_KEY, body.token)
        apply()
      }
    } else {
      removeToken(context)
      throw FailTokenRetrievalException("Failed to get Token")
    }
  }
}
